### ISSUE TITLE
False positive for reportFunctionMemberAccess with recursive functions.

### ISSUE THREAD

**Issue by last-partizan** [microsoft/pyright#6769]
**Describe the bug**

When accessing function member from itself, pyright reports error.

Explicitly adding return type to function fixes the problem (`-> None`  in this case).

**Code or Screenshots**
Here's minimal example:

```python
from functools import lru_cache


@lru_cache
def func():
    func.cache_clear()  # error: Cannot access member "cache_clear" for type "function"
    another_func.cache_clear()


@lru_cache
def another_func():
    func.cache_clear()  # error: Cannot access member "cache_clear" for type "function"
```

**VS Code extension or command-line**

```
> npx pyright@latest --version
pyright 1.1.341
```

**Comment by erictraut** [comment]
In general, return type inference isn't possible for recursive functions. The examples above are not normal recursive functions in that they don't directly invoke themselves, but they effectively create a recursive circularity.

When inferring the return type of `func`, pyright needs to determine whether the code can "fall off the bottom" of the function and return an implicit `None` value. To do this, it needs to determine whether any call expressions invoke functions that do not return (i.e. that return a `NoReturn` type). To do this, it needs to evaluate `func`, which refers to the output of the decorator `@lru_cache`. To evaluate the decorator, the return type of `func` must be known. This is an unresolvable circular dependency. The correct way to handle this, as you've already discovered, is to add an explicit return type annotation to your function.

Pyright's error reporting leaves something to be desired here, but providing a better error would be very difficult. Given that this is an extremely rare edge case and is ultimately a typing bug in the code, I'm not going to invest the time to further improve the error reporting in this case.
